pkg/upgrader/broker/v0.14.0/upgrader.go

```diff
@@ -0,0 +1,124 @@
+	"knative.dev/pkg/apis/duck"
+	kubeclient "knative.dev/pkg/client/injection/kube/client"
+	"knative.dev/pkg/logging"
+)
+
+// Upgrade upgrades all the brokers by applying an Annotation to all the
+// ones that do not have them. This is necessary to ensure that existing Brokers
+// that do not have Annotation will continue to be reconciled by the existing
+// ChannelBasedBroker
+func Upgrade(ctx context.Context) error {
+	logger := logging.FromContext(ctx)
+
+	nsClient := kubeclient.Get(ctx).CoreV1().Namespaces()
+	namespaces, err := nsClient.List(metav1.ListOptions{})
+	if err != nil {
+		logger.Warnf("Failed to list namespaces: %v", err)
+		return err
+	}
+	for _, ns := range namespaces.Items {
+		err = ProcessNamespace(ctx, ns.Name)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func ProcessNamespace(ctx context.Context, ns string) error {
+	logger := logging.FromContext(ctx)
+	logger.Infof("Processing Brokers in namespace: %q", ns)
+
+	eventingClient := eventingclient.Get(ctx)
+	brokerClient := eventingClient.EventingV1alpha1().Brokers(ns)
+	brokers, err := brokerClient.List(metav1.ListOptions{})
+	if err != nil {
+		logger.Warnf("Failed to list brokers for namespace %q: %v", ns, err)
+		return err
+	}
+	for _, broker := range brokers.Items {
+		patch, err := ProcessBroker(ctx, broker)
+		if err != nil {
+			logger.Warnf("Failed to process a Broker \"%s/%s\" : %v", broker.Namespace, broker.Name, err)
+			return err
+		}
+		if len(patch) == 0 {
+			logger.Infof("Broker \"%s/%s\" has annotation already", broker.Namespace, broker.Name)
+			continue
+		}
+
+		// Ok, there are differences, apply the patch
+		logger.Infof("Patching Broker \"%s/%s\" with %q", broker.Namespace, broker.Name, string(patch))
+		patched, err := brokerClient.Patch(broker.Name, types.MergePatchType, patch)
+		if err != nil {
+			logger.Warnf("Failed to patch \"%s/%s\" : %v", broker.Namespace, broker.Name, err)
+		}
+		logger.Infof("Patched \"%s/%s\" successfully new Annotations: %+v", broker.Namespace, broker.Name, patched.ObjectMeta.GetAnnotations())
+	}
+	return nil
+}
+
+// Process a single Broker to see if it needs a patch applied to it or not.
+// Returns non-empty patch bytes if a patch is necessary.
+func ProcessBroker(ctx context.Context, broker v1alpha1.Broker) ([]byte, error) {
+	logger := logging.FromContext(ctx)
+
+	annotations := broker.ObjectMeta.GetAnnotations()
+	if annotations == nil {
+		annotations = make(map[string]string, 1)
+	}
+	if brokerClass, present := annotations[eventing.BrokerClassKey]; present {
+		logger.Infof("Annotation found \"%s/%s\" => %q", broker.Namespace, broker.Name, brokerClass)
+		return []byte{}, nil
+	}
+
+	// No annotations, or missing, add it...
+	modified := broker.DeepCopy()
+	modifiedAnnotations := modified.ObjectMeta.GetAnnotations()
+	if modifiedAnnotations == nil {
+		modifiedAnnotations = make(map[string]string, 1)
+	}
+	if _, present := modifiedAnnotations[eventing.BrokerClassKey]; !present {
```

Comment:
nit: code structure can be simplified if you do `modified := broker.DeepCopy()` at the beginning so you only need to check the modified copy. You waste a copy if nothing needs to be patched but you can avoid checking the annotation twice.